File: windowtool.py
#! /usr/bin/env python
##
##  decorations.py - Remove or show window borders
##  Copyright (C) 2011 Ingo Ruhnke <grumbel@gmail.com>
##
##  This program is free software: you can redistribute it and/or modify
##  it under the terms of the GNU General Public License as published by
##  the Free Software Foundation, either version 3 of the License, or
##  (at your option) any later version.
##
##  This program is distributed in the hope that it will be useful,
##  but WITHOUT ANY WARRANTY; without even the implied warranty of
##  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
##  GNU General Public License for more details.
##
##  You should have received a copy of the GNU General Public License
##  along with this program.  If not, see <http://www.gnu.org/licenses/>.

from gtk.gdk import *
import gtk.gdk
import time
import sys
import argparse

# make it more swiss army knife:
# --move X,Y
# --resize WxH
# --override-redirect False
# --decorations False
# --window WINDOWID
# --keep-above BOOL
# --keep-below BOOL
#
# --windowed to set defaults for window mode
# --fullscreen to set defaults for fullscreen mode
#
# substr=foobar
# title=""
# <integer>

# print "Usage:", sys.argv[0], "show|hide", "WINDOW"
# print "  WINDOW is either a window id as returned by 'xwininfo' or 'wmctrl -l'"
# print "  or the string ACTIVE for the currently focused window."

def str2bool(str):
    return bool(int(str))

class WindowAction(argparse.Action):
    def __call__(self, parser, namespace, values, option_string=None):
        if getattr(namespace, 'override_redirect') == None:
            setattr(namespace, 'override_redirect', False)

        if getattr(namespace, 'decorations') == None:
            setattr(namespace, 'decorations', True)

        if getattr(namespace, 'keep_above') == None:
            setattr(namespace, 'keep_above', False)

        if getattr(namespace, 'keep_below') == None:
            setattr(namespace, 'keep_below', False)

class FullscreenAction(argparse.Action):
    def __call__(self, parser, namespace, values, option_string=None):
        if getattr(namespace, 'override_redirect') == None:
            setattr(namespace, 'override_redirect', True)

        if getattr(namespace, 'decorations') == None:
            setattr(namespace, 'decorations', False)

        if getattr(namespace, 'keep_above') == None:
            setattr(namespace, 'keep_above', True)

        if getattr(namespace, 'keep_below') == None:
            setattr(namespace, 'keep_below', False)

argp = argparse.ArgumentParser(description="Move, resize and set properties on X11 window")

detail_group = argp.add_argument_group("Detail Options")
detail_group.add_argument('-o', '--override-redirect', help="Set override redirect flag",
                          type=str2bool, dest="override_redirect")
detail_group.add_argument('-d', '--decorations', help="Hide or show window decorations",
                          type=str2bool, dest="decorations")
detail_group.add_argument('-a', '--keep-above', help="Keep window above others",
                          type=str2bool, dest="keep_above")
detail_group.add_argument('-b', '--keep-below', help="Keep window below others",
                          type=str2bool, dest="keep_below")
detail_group.add_argument('-m', '--move', dest='position', metavar="X,Y",
                          default=None, type=lambda x: [int(x) for x in x.split(",")],
                          help="Move window to position X,Y (default: None)")
detail_group.add_argument('-r', '--resize', dest='size', metavar="WxH",
                          default=None, type=lambda x: [int(x) for x in x.split("x")],
                          help="Resize window to WxH (default: None)")
detail_group.add_argument('windows', metavar='WINDOWID', type=str, nargs='+',
                          help='Apply actions to window with WINDOWID')

quick_group = argp.add_argument_group("Quick Options")
quick_group.add_argument('-w', '--window', action=WindowAction, nargs=0,
                         help="Short cut for window mode")
quick_group.add_argument('-f', '--fullscreen', action=FullscreenAction, nargs=0,
                         help="Short cut for fullscreen mode")

args = argp.parse_args()

for window in args.windows:
    if window == "ACTIVE":
        w = window_foreign_new((get_default_root_window().property_get("_NET_ACTIVE_WINDOW")[2][0]))
    else:
        w = window_foreign_new(int(window, 16))

    if args.keep_above != None:
        print "Setting keep_above:", args.keep_above
        w.set_keep_above(args.keep_above) # to raise it over the panel

    if args.decorations != None:
        print "Setting decorations:", args.decorations
        if args.decorations:
            time.sleep(1)
            w.set_decorations(gtk.gdk.DECOR_ALL)
        else:
            time.sleep(1)
            w.set_decorations(0)

    if args.override_redirect != None:
        w.set_override_redirect(args.override_redirect)
        gtk.gdk.flush()
        time.sleep(0.5)

    if args.position:
        time.sleep(1)
        w.move(args.position[0], args.position[1])

    elif args.keep_above != None:
        w.set_keep_above(args.keep_above) # to raise it over the panel

    gtk.gdk.flush()

# EOF #

File: windowtool.cpp
#include <charconv>
#include <optional>
#include <string>
#include <string_view>
#include <vector>

#include <fmt/format.h>
#include <gtkmm.h>
#include <gdkmm/window.h>

struct Size
{
  int w;
  int h;
};

struct Position
{
  int x;
  int y;
};

bool parse_bool(std::string const& text)
{
  return bool(std::stoi(text));
}

unsigned int parse_uint(std::string_view text)
{
  unsigned int value;
  std::from_chars_result err;
  if (text.starts_with("0x")) {
    err = std::from_chars(text.data() + 2, text.data() + text.size(), value, 16);
  } else {
    err = std::from_chars(text.data(), text.data() + text.size(), value);
  }

  if (err.ec != std::errc()) {
    throw std::runtime_error(fmt::format("failed to parse: {}", text));
  }

  return value;
}

Position parse_position(std::string text)
{
  Position position;

  if (sscanf(text.c_str(), "%dx%d", &position.x, &position.y) == 2)
  {
    return position;
  } else {
    throw std::runtime_error(fmt::format("failed to convert '{}' to position", text));
  }
}

Size parse_size(std::string text)
{
  Size size;

  if (sscanf(text.c_str(), "%dx%d", &size.w, &size.h) == 2 ||
      sscanf(text.c_str(), "%d,%d", &size.w, &size.h) == 2)
  {
    return size;
  } else {
    throw std::runtime_error(fmt::format("failed to convert '{}' to size", text));
  }
}

struct Options
{
  std::optional<bool> override_redirect = {};
  std::optional<bool> decorations = {};
  std::optional<bool> keep_above = {};
  std::optional<bool> keep_below = {};
  std::optional<Size> size = {};
  std::optional<Position> position = {};
  std::optional<unsigned int> window = {};
  bool verbose = false;
};

void print_help(std::string_view arg0)
{
  fmt::print(
    "Usage: {} [WINDOWID]\n"
    "Manipulate window properties in X11.\n"
    "\n"
    "Options:\n"
    "  --decorations     Enable decorations\n"
    "  --no-decorations  Disable decorations\n"
    "  --keep-above      Enable keep-above\n"
    "  --no-keep-above   Disable keep-above\n"
    "  --keep-below      Enable keep-below\n"
    "  --no-keep-below   Disable keep-below\n"
    "  --move X,Y        Move window to X,Y\n"
    "  --resize XxY      Resize window to WxH\n",
    arg0);
}

Options parse_args(int argc, char** argv)
{
  Options opts;

  for (int i = 1; i < argc; ++i)
  {
    std::string_view arg = argv[i];

    if (!arg.empty() && arg[0] == '-') {
      if (arg == "--help" || arg == "-h") {
        print_help(argv[0]);
        exit(EXIT_SUCCESS);
      } else if (arg == "--override-redirect" || arg == "-o") {
        opts.override_redirect = true;
      } else if (arg == "--no-override-redirect" || arg == "-O") {
        opts.override_redirect = false;
      } else if (arg == "--decorations" || arg == "-d") {
        opts.decorations = true;
      } else if (arg == "--no-decorations" || arg == "-D") {
        opts.decorations = false;
      } else if (arg == "--keep-above" || arg == "-a") {
        opts.keep_above = true;
      } else if (arg == "--no-keep-above" || arg == "-A") {
        opts.keep_above = false;
      } else if (arg == "--keep-below" || arg == "-b") {
        opts.keep_below = true;
      } else if (arg == "--no-keep-below" || arg == "-B") {
        opts.keep_below = false;
      } else if (arg == "--move" || arg == "-m") {
        i += 1;
        opts.position = parse_position(std::string(argv[i]));
      } else if (arg == "--resize" || arg == "-r") {
        i += 1;
        opts.size = parse_size(std::string(argv[i]));
      } else {
        throw std::runtime_error(fmt::format("unknown argument: {}", arg));
      }
    } else {
      opts.window = parse_uint(arg);
    }
  }

  return opts;
}

int main(int argc, char** argv)
{
  Options opts = parse_args(argc, argv);

  Gtk::Main kit(argc, argv);

  Glib::RefPtr<Gdk::Window> w;
  if (opts.window) {
    // can't find foreign_new() in Gtkmm
    fmt::print("using window {}\n", *opts.window);
    GdkWindow* win = gdk_window_foreign_new(*opts.window);
    Glib::RefPtr<Gdk::Drawable> drawable(Glib::wrap(win));
    w = Glib::RefPtr<Gdk::Window>::cast_dynamic(drawable);
  } else {
    fmt::print("using active window\n");
    w = Gdk::Screen::get_default()->get_active_window();
  }

  if (!w) {
    throw std::runtime_error(fmt::format("invalid window id: {}", *opts.window));
  }

  if (opts.keep_above) {
    fmt::print("Setting keep_above: {}\n", *opts.keep_above);
    w->set_keep_above(*opts.keep_above);
  }

  if (opts.decorations) {
    fmt::print("Setting decorations: {}\n", *opts.decorations);

    if (*opts.decorations) {
      Glib::usleep(1000000);
      w->set_decorations(Gdk::DECOR_ALL);
    } else {
      Glib::usleep(1000000);
      w->set_decorations(Gdk::WMDecoration{});
    }
  }

  if (opts.override_redirect) {
    w->set_override_redirect(*opts.override_redirect);
    Gdk::flush();
    Glib::usleep(500000);
  }

  if (opts.position) {
    Glib::usleep(1000000);
    w->move(opts.position->x, opts.position->y);
  }

  if (opts.size) {
    Glib::usleep(1000000);
    w->resize(opts.size->w, opts.size->h);
  }

  if (opts.keep_below) {
    fmt::print("Setting keep_below: {}\n", *opts.keep_below);
    w->set_keep_below(*opts.keep_below);
  }

  fmt::print("flush\n");
  Gdk::flush();

  return 0;
}

/* EOF */
